pallets/ajuna-affiliate/src/lib.rs

```diff
@@ -0,0 +1,269 @@
+// Ajuna Node
+// Copyright (C) 2022 BlogaTech AG
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+pub mod traits;
+
+use frame_support::pallet_prelude::*;
+
+use traits::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use sp_runtime::ArithmeticError;
+
+	pub type AffiliatedAccountsOf<T, I> =
+		BoundedVec<<T as frame_system::Config>::AccountId, <T as Config<I>>::AffiliateMaxLevel>;
+
+	pub type AccountIdFor<T> = <T as frame_system::Config>::AccountId;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self, I>>
+			+ IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The rule identifier type at runtime.
+		type RuleIdentifier: Parameter + MaxEncodedLen;
+
+		/// The rule type at runtime.
+		type RuntimeRule: Parameter + MaxEncodedLen;
```

Comment:
Yes, exactly, this is how I meant it. I have been thinking some more about a trait bound we could introduce here, like `ExecuteRule`. However, as this pallet has no predefined hooks, meaning the pallet that uses the affiliate pallet can decide when to use this it won't be needed.

However, then that pallet needs to invoke the execution of the rule itself, then the question is, why do we even store the rule here? It is not needed, the invoking pallet already knows, which rule it wants to execute. Then this pallet, would actually only need to handle affiliate relations and no rules at all, and the other pallet simply wants to check if an account is an affiliator or not.

Alternatively, if we really and this to be generic, then maybe the `RuleId` does not correspond to the extrinsic enum variant. This would allow some runtime to even have two pallets, one that wants to add affiliate relations, and another one that would want to execute the rule. In this case, we'd need to store the rule here, but we'd need to have a trait bound so that the other pallet could execute the rule, if it finds one.

This is probably one more fundamental thing we have to discuss in the brainstorming session tomorrow.